enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,253 @@
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster breaking configuration mistakes.
+* 
+### Non-Goals
+
+* Replace SRIOV operator
+
+## Proposal
+
+#### Option A
+
+The node networking configuration shall driven by MCO (and by extension kubernetes-nmstate) and executed by
+[nmstate](https://nmstate.github.io/).
+
+Do a KubeletConfig like integration:
+
+* NodeNetworkState will be introduced. This allows the user to see the network state of a node in the same declarative language that can be used to configure it.
+* The user will configure with NodeNetworkConfigurationPolicy (NNCP). This is a declaritive method of defining network state, to be applied by nmstate.
+* NNCP, just like KubeletConfig, targets a MachinePool
+* The MachineConfigDaemon (**MCD**) vendors the knmstate configuring/rollback functionality.
+* MCD runs a vendored subprocess that keeps NodeNetworkState (**NNS**) up to date with the OpenShift API
+* NMState is invoked with Podman (Python project with dependencies outside of stdlib)
+
+1. User Creates **NNCP**
+2. KNMStateController takes the **NNCP** desiredConfig and puts it as a **MachineConfig** (MC) for the desired pool
+3. MachineConfigController renders a new config and its update subcontroller starts updating desiredState of the Nodes
+4. MachineConfigDaemon sees the diff between currentState and desiredState
+5. In the case all the new rendered MC is equal except the **networkState** section, instead of reboot 
+    * MCD writes networkState as is (nmstate valid yaml) and writes it to /tmp
+    * MCD calls nmstatectl set --no-commit /tmp/uuid-of-the-desiredState.yaml
+    * If no connectivity loss, MCD calls nmstatectl commit [PREVIOUS CHECKPOINT]
+    * Otherwise, MCD calls nmstatectl rollback
+* MCD sets the currentState to match the desiredState.
+
+#### Option B
+
+Interact with kubernetes-nmstate handler through MCD. Prepare hooks in MCD for that. User would create MCs, those would be propagated to MCD, MCD would delegate to kubernetes-nmstate handler.
+
+#### Recommendation
+
+We believe that option A poses the best combination of using existing MCO with the introduction of kubernetes-nmstate.
+
+### User Stories
+
+#### Host network configuration (Telco)
+
+As an OpenShift administrator working in Telco, I need to add and remove VLANs on day-2+, more often than the update cycle. Reboot of the whole baremetal cluster could take days and/or impact capacity.
+
+#### L2 networking for VMs (CNV)
+
+As an OpenShift administrator using Container Native Virtualization, I need to configure ad-hoc VLANs and Linux bridges on day-2. These bridges are used to provide L2 connectivity to VMs. 
+
+#### L2 networking for Pods (Telco, CNF)
+
+As an OpenShift administrator working in Telco, I need to configure bondings and ad-hoc VLANs on day-2. Macvlan ifaces will be attached to these vlans and plugged into containers.
```

Comment:
This talks about "Telco" but nothing about it seems telco-related. I think this is really "We have a specific customer who says they want to do X"

Reply:
I'm not really sure how to respond here. Does it matter? The need for the feature doesn't change.

Reply:
My complaint in general is that the user stories aren't really adding anything. They're all basically just "As a user, I will use the feature described above in the Proposal section".

But it looks like this is pretty standard for openshift enhancements I guess. Don't mind me.